pyccel/parser/syntax/openmp.py

```diff
@@ -31,142 +35,547 @@ def __init__(self, **kwargs):
+
+        return OMP_Cancel_Construct(self.txt)
+
+class OmpTargetConstruct(BasicStmt):
+    """Class representing a Target construct."""
+    def __init__(self, **kwargs):
+        """
+        """
+        self.clauses  = kwargs.pop('clauses')
+        self.name     = kwargs.pop('name')
+        self.combined = kwargs.pop('combined', None)
+
+        _valid_clauses = (OmpPrivate,
+                          OmpLastPrivate,
+                          OmpinReduction,
+                          OmpDepend,
+                          OmpMap)
+
+        self.com = None
+        if isinstance(self.combined, (OmpTargetParallel, OmpTargetTeams)):
+            self.com = self.combined.expr
+            if 'distribute' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_Distribute_clauses
+            if 'parallel' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_parallel_clauses
+            if 'simd' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_simd_clauses
+            if 'for' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_loop_clauses
+            if 'teams' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_teams_clauses
+
+        self.txt = ''
         for clause in self.clauses:
             if isinstance(clause, _valid_clauses):
-                txt = '{0} {1}'.format(txt, clause.expr)
+                if isinstance(clause, OmpCopyin) and isinstance(self.combined, OmpTargetParallel):
+                    raise TypeError('Wrong clause for OmpTargetConstruct')
+                self.txt = '{0} {1}'.format(txt, clause.expr)
             else:
-                raise TypeError('Wrong clause for OmpSingleConstruct')
+                raise TypeError('Wrong clause for OmpTargetConstruct')
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        if DEBUG:
+            print("> OmpTargetConstruct: expr")
+
+        return OMP_Target_Construct(self.txt, self.com)
 
-        return OMP_Single_Construct(txt)
+class OmpTeamsConstruct(BasicStmt):
+    """Class representing a Teams construct."""
+    def __init__(self, **kwargs):
+        """
+        """
+        self.name     = kwargs.pop('name')
+        self.clauses  = kwargs.pop('clauses')
+        self.combined = kwargs.pop('combined', None)
+
+        _valid_clauses = _valid_teams_clauses
+
+        self.com = None
+        if isinstance(self.combined, OmpDistributeCombined):
+            self.com = self.combined.expr
+            _valid_clauses = _valid_clauses + _valid_Distribute_clauses
+            if 'simd' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_simd_clauses
+            if 'parallel' in self.combined.expr:
+                _valid_clauses = _valid_clauses + _valid_parallel_clauses
+                _valid_clauses = _valid_clauses + _valid_loop_clauses
```

Comment:
I am surprised there are not more statements like this across your different classes. Is each combined type (e.g. `OmpDistributeCombined`, `OmpForSimd`) only associated with one construct type (e.g. `OmpTeamsConstruct`) ?
If not then I wonder if it makes sense for combined types to have a `valid_clauses` method which returns the additional clauses which are valid for that construct? That way you won't have to copy and paste if blocks if a combined type can be used by multiple constructs?

Reply:
i agree with you, but here each combined type is associated to only one construct type, but two combined types can have the same one or more elements, I'm gonna try to simplify it